canister/src/blocktree.rs

```diff
@@ -384,4 +415,118 @@ mod test {
         // of the root and child with the greatest difficulty which is 5 + 10 = 15.
         assert_eq!(difficulty_based_depth(&block_tree, Network::Mainnet), 15);
     }
+
+    #[test]
+    fn test_blocks_with_depths_separated_by_heights_only_root() {
+        let genesis_block = BlockBuilder::genesis().build();
+        let block_tree = BlockTree::new(genesis_block.clone());
+        let blocks_with_depths_separated_by_heights =
+            block_tree.blocks_with_depths_separated_by_heights();
+
+        // The number of rows in blocks_with_depths_separated_by_heights should be 1.
+        // The row should have only 1 column.
+        assert_eq!(blocks_with_depths_separated_by_heights.len(), 1);
+        assert_eq!(blocks_with_depths_separated_by_heights[0].len(), 1);
+
+        let (block, depth) = blocks_with_depths_separated_by_heights[0][0];
+        // Depth of the genesis block should be 1.
+        assert_eq!(block.block_hash(), genesis_block.block_hash());
+        assert_eq!(depth, 1);
+    }
+
+    #[test]
+    fn test_blocks_with_depths_separated_by_heights_chain() {
+        let mut blocks = vec![BlockBuilder::genesis().build()];
+        let chain_len = 10;
+        for i in 1..chain_len {
+            blocks.push(BlockBuilder::with_prev_header(blocks[i - 1].header()).build())
+        }
```

Comment:
You may find it simpler to use the `BlockchainBuilder`:

```suggestion
        let chain_len = 10;
        let blocks = let chain = BlockChainBuilder::new(chain_len).build();
```